src/ui/win32/RichPresenceDialog.cpp

```diff
@@ -0,0 +1,113 @@
+#include "RichPresenceDialog.hh"
+
+#include "RA_Core.h"
+#include "RA_Resource.h"
+#include "RA_RichPresence.h"
+
+namespace ra {
+namespace ui {
+namespace win32 {
+
+bool RichPresenceDialog::Presenter::IsSupported(const ra::ui::WindowViewModelBase& oViewModel)
+{
+    return (dynamic_cast<const ra::ui::viewmodels::RichPresenceMonitorViewModel*>(&oViewModel) != nullptr);
+}
+
+void RichPresenceDialog::Presenter::ShowModal(ra::ui::WindowViewModelBase& oViewModel)
+{
+    ShowWindow(oViewModel);
+}
+
+void RichPresenceDialog::Presenter::ShowWindow(ra::ui::WindowViewModelBase& oViewModel)
+{
+    auto& oRichPresenceViewModel = reinterpret_cast<ra::ui::viewmodels::RichPresenceMonitorViewModel&>(oViewModel);
+
+    if (m_pDialog == nullptr)
+    {
+        m_pDialog.reset(new RichPresenceDialog(oRichPresenceViewModel));
+        m_pDialog->CreateDialogWindow(MAKEINTRESOURCE(IDD_RA_RICHPRESENCE), this);
+    }
+    
+    if (m_pDialog->ShowDialogWindow())
+        m_pDialog->StartTimer();
+}
+
+void RichPresenceDialog::Presenter::OnClosed()
+{
+    m_pDialog.reset();
+}
+
+// ------------------------------------
+
+RichPresenceDialog::RichPresenceDialog(ra::ui::viewmodels::RichPresenceMonitorViewModel& vmRichPresenceDisplay)
+    : DialogBase(vmRichPresenceDisplay)
+{
+    m_hFont = CreateFont(15, 0, 0, 0, FW_NORMAL, FALSE, FALSE, FALSE, DEFAULT_CHARSET, OUT_DEFAULT_PRECIS, CLIP_DEFAULT_PRECIS, CLEARTYPE_QUALITY, VARIABLE_PITCH, nullptr);
+
+    m_bindWindow.SetInitialPosition(ra::ui::win32::bindings::RelativePosition::After, ra::ui::win32::bindings::RelativePosition::Far, "Rich Presence Monitor");
+    m_bindWindow.BindLabel(IDC_RA_RICHPRESENCERESULTTEXT, ra::ui::viewmodels::RichPresenceMonitorViewModel::DisplayStringProperty);
+}
+
+RichPresenceDialog::~RichPresenceDialog()
+{
+    DeleteFont(m_hFont);
+}
+
+static void UpdateDisplayString(ra::ui::WindowViewModelBase& vmWindow)
+{
+    auto& vmRichPresence = dynamic_cast<ra::ui::viewmodels::RichPresenceMonitorViewModel&>(vmWindow);
+    vmRichPresence.UpdateDisplayString();
+}
+
+void RichPresenceDialog::OnInitDialog()
+{
+    DialogBase::OnInitDialog();
+
+    SendMessage(GetDlgItem(GetHWND(), IDC_RA_RICHPRESENCERESULTTEXT), WM_SETFONT, (WPARAM)m_hFont, LPARAM{});
+
+    UpdateDisplayString(m_vmWindow);
+}
+
+void RichPresenceDialog::OnDestroy()
+{
+    StopTimer();
+
+    DialogBase::OnDestroy();
+}
+
+INT_PTR CALLBACK RichPresenceDialog::DialogProc(HWND hDlg, UINT nMsg, WPARAM wParam, LPARAM lParam)
+{
+    switch (nMsg)
+    {
+        case WM_TIMER:
```

Comment:
`WM_TIMER `message cracker for future reference, if there's only timer shouldn't be `TimerProc`?
`void OnTimer(HWND hwnd, UINT id)`

Reply:
The timer code is temporary. The plan is to move this to the RichPresenceInterpretter using something along these lines:
```
ITimerService.Schedule(std::chrono::minutes(2), UpdateDisplayText);
```
where UpdateDisplayText schedules another callback for itself two minutes later.